### PR TITLE
Handle unread body from request

### PR DESCRIPTION
**Pull request by bartlomieju**
Follow up #22 #49.

This PR provides most basic implementation that handles unread request body without reading all data to memory. As discussed in #49 it follows Golang's implementation.

### CODE REVIEW

File: http/server.ts [hunk 1]
```diff
@@ -271,6 +340,10 @@ export class ServerRequest {
     }
 
     await this.w.flush();
+
+    if (!this._shouldReuseConnection()) {
+      this.conn.close();
```
Comment (bartlomieju) [inline http/server.ts]: @ry you said that [`closeWriteAndWait`](https://github.com/golang/go/blob/9e277f7d554455e16ba3762541c53e9bfc1d8188/src/net/http/server.go#L1688) is what we need. I don't fully understand it's purpose. Could you follow up?

**Comment by bartlomieju** [comment]
I'll wait on #188 before finishing this

**Comment by bartlomieju** [comment]
I've tried to rebase my branch against master, but given latest changes in #188 it's pretty hard.

Main loop in `serve` has:
```ts
for (const { req, conn } of queueToProcess) {
      if (req) {
        const res = createResponder(conn);
        yield { req, res };
      }
      serveConn(env, conn);
}
```
request and response are completely decoupled, which gives no easy way to check if request's body was fully read before responding, which leaves #49 unresolved

Maybe we could talk about rewriting it to something like this:
```
for (const { req, conn } of queueToProcess) {
      if (req) {
        // iterator should return response
        const res = yield req;

        // close connection if request body 
        // hasn't been consumed
        writeResponse(res, req);
      }

      // check if connection can be reused
      // get closed() should be added to `Conn` 
      if (!conn.closed) {
        serveConn(env, conn);
      }
}
```

CC @ry @kevinkassimo

**Comment by keroxp** [comment]
@bartlomieju 

Isn't  code like this enough? After #188, `req.body` is one of instance of `BodyReader` or `ChunkedBodyReader`. It can consume body to exactly end of the stream. If body was consumed entirely by user, no additional io operations happen.

```ts
for (const { req, conn } of queueToProcess) {
  if (req) {
    const res = createResponder(conn);
    yield { req, res };
    await copy({
      write: async p => p.byteLength
    }, req.body)
  }
  serveConn(env, conn);
}
```

**Comment by bartlomieju** [comment]
@keroxp agreed, however logic deciding whether or not it should be consumed or connection closed is much more complicated. We wanted to base it on Golang's implementation, here's old PR with discussion for context: #49

**Pull request closed without merging**